src/v/kafka/server/snc_quota_manager.h

```diff
@@ -19,10 +19,84 @@
 #include <seastar/core/sharded.hh>
 
 #include <chrono>
+#include <functional>
 #include <optional>
+#include <vector>
 
 namespace kafka {
 
+/// Represents a homogenous pair of values that correspond to
+/// ingress and egress side of the same entity
+template<class T>
+struct inoutpair {
+    T in;
+    T out;
+};
+
+template<>
+struct inoutpair<void> {};
+
+namespace detail {
+
+template<class T>
+struct unwrap_inoutpair {};
+
+template<class U>
+struct unwrap_inoutpair<inoutpair<U>> {
+    using type = U;
+};
+
+template<class U>
+struct unwrap_inoutpair<const inoutpair<U>> {
+    using type = const U;
+};
+
+template<class U>
+struct unwrap_inoutpair<inoutpair<U>&> {
+    using type = U&;
+};
+
+template<class U>
+struct unwrap_inoutpair<const inoutpair<U>&> {
+    using type = const U&;
+};
+
+template<class T>
+using unwrap_inoutpair_t = typename unwrap_inoutpair<T>::type;
+
+} // namespace detail
+
+/// Applies \p f to a set of values wrapped in \ref inoutpair, once for ingress
+/// side and another time for egress side. The result of \p f is returned
+/// wrapped into \ref inoutpair too. void return type supported.
+template<
+  class... T,
+  std::invocable<detail::unwrap_inoutpair_t<T>...> F,
+  class TRes = std::invoke_result_t<F, detail::unwrap_inoutpair_t<T>...>>
+inoutpair<TRes> to_each(F f, T&&... args) {
+    if constexpr (std::is_void_v<TRes>) {
+        std::invoke(f, args.in...);
+        std::invoke(f, args.out...);
+        return {};
+    } else {
+        return inoutpair<TRes>{
+          .in = std::invoke(f, args.in...), .out = std::invoke(f, args.out...)};
+    }
+}
+
+/// Copies vector of inoutpairs to an inoutpair of vectors
+template<class T>
+inoutpair<std::vector<T>> to_inside_out(const std::vector<inoutpair<T>>& v) {
```

Comment:
there is only one caller to this it might make sense to perform this inline where it's called because the name isn't very intuitive about what is happening, but the actual implementation makes it clear.

Reply:
kept the function but moved it to the .cc